soroban-env-host/src/crypto/poseidon/poseidon_params.rs

```diff
@@ -0,0 +1,123 @@
+use super::super::metered_scalar::MeteredScalar;
+use super::INVALID_INPUT;
+use crate::{crypto::poseidon::SUPPORTED_SBOX_DEGREES, ErrorHandler, Host, HostError, Val};
+
+#[derive(Clone, Debug)]
+pub struct PoseidonParams<S: MeteredScalar> {
+    pub(crate) t: usize, // statesize
+    pub(crate) d: usize, // sbox degree
+    pub(crate) rounds_f_beginning: usize,
+    pub(crate) rounds_p: usize,
+    pub(crate) rounds: usize,
+    pub(crate) mds: Vec<Vec<S>>,
+    pub(crate) round_constants: Vec<Vec<S>>,
+}
+
+impl<S: MeteredScalar> PoseidonParams<S> {
+    /// Creates new PoseidonParams with full validation using Host for error reporting.
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
+        host: &Host,
+        t: u32,
+        d: u32,
+        rounds_f: u32,
+        rounds_p: u32,
+        mds: Vec<Vec<S>>,
+        round_constants: Vec<Vec<S>>,
+    ) -> Result<Self, HostError> {
+        if !SUPPORTED_SBOX_DEGREES.contains(&d) {
+            return Err(host.error(
+                INVALID_INPUT,
+                "Unsupported s-box degree",
```

Comment:
nit: Shall we prefix all the errors with `Poseidon: ` and `Poseidon2: ` where possible for clarity?